features/310-prerender-gear/prerender.md

```diff
@@ -0,0 +1,105 @@
+
+Response headers are then modified. Specifically, `content-length` and `content-encoding` will be removed.
+
+Finally, the modified DOM and headers are put into prerendered html cache using the forwarded url as keys before being sent back to gateway. A configurable expiry time is also applied to the cache such that they expire automatically.
+
+#### Redirect
+
+Note that since the forwarded url is used directly as the cache key, even if the forwarded url in a future request is the same as any of the redirected urls prerender has encountered during navigation, the request is a cache miss. For example:
+
+1. gateway requests url `http://a.com`
+2. prerender sees a caches miss, proceed to navigates to `http://a.com`
+3. prerender is redirected to `http://b.com`
+4. prerender captures the DOM of `http://b.com`, then stores it with key `http://a.com` and serves it
+5. gateway requests url `http://b.com`
+6. prerender sees a cache miss (it only has cache for `http://a.com`), proceeds to navigates to `http://b.com`
+7. prerender captures the DOM of `http://b.com` (again), then stores it with key `http://b.com` and serves it
+
+Notice how in (6) prerender consider `http://b.com` a cache miss even though it has been redirected to it.
+
+### Invalidating cache
+
+Since prerender supports some configuration about cache, when a configuration is changed such that according to the new configuration old cache is no longer valid, it is necessary to manually reset prerendered html cache. Prerender supports invalidation via a `origin` and `path` protocol.
+
+## API
+
+### Configuration
+
+`seo-prerender`: Boolean. `true` to enable prerender, `false` to disable.
+
+`seo-prerender-expiry`: Time. The amount of time a prerendrered html is kept in cache.
+
+`seo-prerender-capture-wait`: Time. The amount of time prerender waits before capturing DOM such that javascript, etc can have time to run.
+
+`seo-prerender-user-agent`: Comma separated strings. `user-agent` that is considered a search engine.
+
+### Endpoint
+
+#### Prerender
+
+Path:
+
+`/`
+
+Query parameters:
+
+- `u`: The url to render. A url with slash at the end and a url without slash at the end are cached separately.
+- `expiry`: Corresponds to `seo-prerender-expiry` in configuration, with the time specified in seconds. i.e. To have an expiry of one hour one should pass `3600`. Optional.
+- `renderWait`: Corresponds to `seo-prerender-capture-wait` in configuration, with the time specified in seconds. i.e. To wait 30 seconds one should pass `30`. Optional.
+
+Response:
+
+- `200`: If `content-type` of the original response of the url is not `text/html`, the original response and headers, modified DOM and headers otherwise.
```

Comment:
You may want to have a look on https://golang.org/src/net/http/httputil/reverseproxy.go?s=5699:5774#L188 to know how to proxy a request correctly. Preferably we can use that directly to proxy the request, and continue our actual work on the response if the `content-type` is `text/html`.

Reply:
It works when the response is not an html, but when the response is html we have to make the headless browser navigate to that url again (which is redundant) since there is no way to just let the headless browser use the received response.

Reply:
So we cannot just give an HTML to the headless browser?

Reply:
There was a method exposed from chrome dev tools (the underlying protocol `chromedp` is using), but there are required parameters that I do not know how to fetch. The method in question in chrome dev tool is [here](https://chromedevtools.github.io/devtools-protocol/tot/Page#method-setDocumentContent), whereas the binding in `chromedp` is [here](https://godoc.org/github.com/chromedp/cdproto/page#SetDocumentContent). Essentially, there was no way to get a proper frame id without first navigating somewhere. Maybe `0` would work...Not sure.

And can you tell me why you want to use a "real" proxy server? I'm not familiar with proxy in general but my gut feeling is it seems like an overkill. Well maybe if there was a clear and easy way to set html content in the first place I would have gone the proxy way...

Reply:
If you look at the source code I posted, proxying a request robustly involves many details. So I prefer we just leave this work to a existing solution instead of rolling our own one. Anyway, we can always improve (or fix) the implementation later on. It is not a problem that we need to deal with now.

Reply:
I did look at it, and there are more lines of code than my `main.go` lol. I get your point, and my reply is that maybe we don't need all the protocol change detection, rerouting, adding forwarded headers, etc. Mainly because I think of the operation as "inspecting the browser and return whatever is useful" rather than "proxy the request then process response using browser". I will admit I am not sure if my implementation is breaking any part of the standard though. Thanks for the input.